File: backend/src/main/scala/order/api/RejectOrderApi.scala
package order.api

import domain.shared.given

import cats.effect.IO
import domain.order.RejectOrderRequest
import domain.shared.{DeliveryAppState, OrderId, UserRole}
import order.app.rejectOrder
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.io.*
import shared.api.routing.*
import shared.app.*

val rejectOrderApi: FixedMethodApi1[OrderId, DeliveryAppState] =
  jsonPostApi1[OrderId, RejectOrderRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("orders")),
    List(routeSegment("reject")),
  )

File: frontend/src/apis/order/RejectOrderApi.ts
import type { DeliveryAppState, OrderId, RejectOrderRequest } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath1, defineJsonPostApi1, routeSegment } from '@/system/api/TypedApiDefinitions'

export const rejectOrderApiDefinition = defineJsonPostApi1<OrderId, RejectOrderRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('orders')],
    [routeSegment('reject')],
  )

export function rejectOrder(orderId: OrderId, payload: RejectOrderRequest) {
  return postNormalizedDeliveryState(buildApiPath1(rejectOrderApiDefinition, orderId), payload)
}
